RecoTracker/TkSeedGenerator/plugins/JetCoreMCtruthSeedGenerator.cc

```diff
@@ -0,0 +1,580 @@
+  return clustDirs;
+}
+
+std::vector<PSimHit> JetCoreMCtruthSeedGenerator::coreHitsFilling(edm::Handle<std::vector<PSimHit>> simhits,
+                                                                  const GeomDet* globDet,
+                                                                  GlobalVector bigClustDir,
+                                                                  const reco::Vertex& jetVertex) {
+  std::vector<PSimHit> goodSimHit;
+  std::vector<PSimHit>::const_iterator shIt = simhits->begin();
+  for (; shIt != simhits->end(); shIt++) {  //loop deset
+    const GeomDet* det = geometry_->idToDet((*shIt).detUnitId());
+    if (det != globDet)
+      continue;
+    std::pair<bool, Basic3DVector<float>> interPair =
+        findIntersection(bigClustDir, (reco::Candidate::Point)jetVertex.position(), det);
+    if (interPair.first == false)
+      continue;
+    Basic3DVector<float> inter = interPair.second;
+    auto localInter = det->specificSurface().toLocal((GlobalPoint)inter);
+
+    if (std::abs(((*shIt).localPosition()).x() - localInter.x()) / pitchX <= jetDimX / 2 &&
+        std::abs(((*shIt).localPosition()).y() - localInter.y()) / pitchY <= jetDimY / 2) {
+      goodSimHit.push_back((*shIt));
+    }
+  }
+  return goodSimHit;
+}
+
+std::pair<std::vector<SimTrack>, std::vector<SimVertex>> JetCoreMCtruthSeedGenerator::coreTracksFilling(
+    std::vector<PSimHit> goodSimHit,
+    const std::vector<SimTrack>* simtracksVector,
+    const std::vector<SimVertex>* simvertexVector) {
+  std::vector<SimTrack> goodSimTrk;
+  std::vector<SimVertex> goodSimVtx;
+
+  for (uint j = 0; j < simtracksVector->size(); j++) {
+    for (std::vector<PSimHit>::const_iterator it = goodSimHit.begin(); it != goodSimHit.end(); ++it) {
+      SimTrack st = simtracksVector->at(j);
+      if (st.trackId() == (*it).trackId()) {
+        for (uint v = 0; v < simvertexVector->size(); v++) {
+          SimVertex sv = simvertexVector->at(v);
+          if ((int)sv.vertexId() == (int)st.vertIndex()) {
+            goodSimTrk.push_back(st);
+            goodSimVtx.push_back(sv);
+          }
+        }
+      }
+    }
+  }
+  std::pair<std::vector<SimTrack>, std::vector<SimVertex>> output(goodSimTrk, goodSimVtx);
+  return output;
+}
+
+std::pair<std::vector<SimTrack>, std::vector<SimVertex>> JetCoreMCtruthSeedGenerator::coreTracksFillingDeltaR(
+    const std::vector<SimTrack>* simtracksVector,
+    const std::vector<SimVertex>* simvertexVector,
+    const GeomDet* globDet,
+    const reco::Candidate& jet,
+    const reco::Vertex& jetVertex) {
+  std::vector<SimTrack> goodSimTrk;
+  std::vector<SimVertex> goodSimVtx;
+
+  GlobalVector jetDir(jet.px(), jet.py(), jet.pz());
+
+  for (uint j = 0; j < simtracksVector->size(); j++) {
+    SimTrack st = simtracksVector->at(j);
+    GlobalVector trkDir(st.momentum().Px(), st.momentum().Py(), st.momentum().Pz());
+    if (Geom::deltaR(jetDir, trkDir) < deltaR_) {
+      if (st.charge() == 0)
+        continue;
```

Comment:
looks like this can be outside the deltaR to save some computation

Reply:
ok